aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,179 @@
+      dst_usr_dims;
+  dnnl::memory::dims m1_usr_strides, m2_usr_strides, scale_usr_strides,
+      zp_usr_strides, dst_usr_strides;
+  int compressed_k = (int)(k / 8);
+  int num_groups = (int)(k / group_size);
+  m1_usr_dims = {m, k};
+  m1_usr_strides = {m1.stride(0), m1.stride(1)};
+  m2_usr_dims = {compressed_k, n};
+  m2_usr_strides = {1, compressed_k}; // k dim contiguous, 4bit pack into s32
+
+  scale_usr_dims = {num_groups, n};
+  scale_usr_strides = {n, 1};
+  zp_usr_dims = {num_groups, n};
+  zp_usr_strides = {n, 1};
+  dst_usr_dims = {m, n};
+  dst_usr_strides = {dst.stride(0), dst.stride(1)};
+
+  dnnl::memory::desc m1_usr_md, m2_usr_md, scale_usr_md, zp_usr_md, dst_usr_md;
+
+  m1_usr_md = dnnl::memory::desc(m1_usr_dims, m1_usr_dt, m1_usr_strides);
+  m2_usr_md = dnnl::memory::desc(m2_usr_dims, m2_usr_dt, m2_usr_strides);
+  scale_usr_md =
+      dnnl::memory::desc(scale_usr_dims, scale_usr_dt, scale_usr_strides);
+  zp_usr_md = dnnl::memory::desc(zp_usr_dims, zp_usr_dt, zp_usr_strides);
+  dst_usr_md = dnnl::memory::desc(dst_usr_dims, dst_usr_dt, dst_usr_strides);
+
+  // create usr memory
+  auto dst_usr_m = make_onednn_memory(dst_usr_md, engine, dst.data_ptr());
+  auto scale_usr_m = make_onednn_memory(scale_usr_md, engine, scale.data_ptr());
+  auto zp_usr_m = make_onednn_memory(zp_usr_md, engine, zp.data_ptr());
+
+  // Construct md for primitive creation
+  // The xxx_md describes what kinds of matmul the oneDNN does.
+  // The problem for this op is [m, k] x [k, n] => [m, n] matmul.
+  auto m1_dt = m1_usr_dt; // bf16
+  // Tell oneDNN the weight dtype we want manipulate is u4,
+  // library needs infer how to unpack u4 data based on the m2_usr_md (s32).
+  auto m2_dt = dnnl::memory::data_type::u4;
+  auto scale_dt = scale_usr_dt; // bf16
+  // Tell oneDNN the zp dtype we want manipulate is s8
+  // library needs infer how to unpack s8 data based on the m2_usr_md.
+  auto zp_dt = zp_usr_dt; // should be s8, currently
+  auto dst_dt = dst_usr_dt;
+
+  dnnl::memory::desc m1_md, m2_md, scale_md, zp_md, dst_md;
+  dnnl::memory::dims m1_dims, m2_dims, scale_dims, zp_dims, dst_dims;
+  dnnl::memory::dims m1_strides, m2_strides, scale_strides, zp_strides,
+      dst_strides;
+
+  m1_dims = m1_usr_dims; // {m, k}
+  m1_strides = m1_usr_strides; // {k, 1}
+  m2_dims = {k, n};
+  m2_strides = {n, 1};
+  scale_dims = scale_usr_dims; // {k//group_size, n}
+  scale_strides = scale_usr_strides;
+  zp_dims = zp_usr_dims;
+  zp_strides = zp_usr_strides;
+  dst_dims = dst_usr_dims;
+  dst_strides = dst_usr_strides;
+
+  m1_md = dnnl::memory::desc(m1_dims, m1_dt, m1_strides);
+  m2_md = dnnl::memory::desc(m2_dims, m2_dt, m2_strides);
+  scale_md = dnnl::memory::desc(scale_dims, scale_dt, scale_strides);
+  zp_md = dnnl::memory::desc(zp_dims, zp_dt, zp_strides);
+  dst_md = dnnl::memory::desc(dst_dims, dst_dt, dst_strides);
+
+  std::unordered_map<int, dnnl::memory> args;
+  dnnl::post_ops po = attr.extract_post_ops(dst);
```

Comment:
The `po` should be useless. Has this file been added to torch linter?

Reply:
* Post-ops is not required at present. We can remove the post op and added it back when it is necessary.
*  All file in `xpu/detail/*.cpp` is in linter. I met this before. It should caused that, linter does not check this kind of coding style issue.